### PR TITLE
fix exprt::opX accesses in simplifier

### PR DESCRIPTION
**Pull request by kroening**
This improves type safety.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e8cd19214ea7a435a6005714a308070caa708b21).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121876996

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -553,7 +553,7 @@ simplify_exprt::simplify_typecast(const typecast_exprt &expr)
       to_bitvector_type(expr_type).get_width())
   {
     auto new_expr = expr;
-    new_expr.op() = expr.op().op0();
+    new_expr.op() = to_typecast_expr(expr.op()).op();
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp:555]: If we're doing this (which I think is a good idea) then we should get rid of the `expr.op().operands().size() == 1 &&` in the condition.

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -637,7 +637,7 @@ simplify_exprt::simplify_typecast(const typecast_exprt &expr)
     op_type.id() == ID_pointer && expr.op().operands().size() == 1)
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: We should get rid of ` && expr.op().operands().size() == 1`

File: src/util/simplify_expr_array.cpp [hunk 3]
```diff
@@ -180,7 +181,7 @@ simplify_exprt::simplify_index(const index_exprt &expr)
 
       // add offset to index
       mult_exprt offset(from_integer(*sub_size, array.op1().type()), index);
-      plus_exprt final_offset(array.op1(), offset);
+      plus_exprt final_offset(to_byte_extract_expr(array).offset(), offset);
```
Comment (tautschnig) [inline src/util/simplify_expr_array.cpp]: Should also be fixed in the line above.

File: src/util/simplify_expr_struct.cpp [hunk 4]
```diff
@@ -151,12 +151,13 @@ simplify_exprt::simplify_member(const member_exprt &expr)
...
 
-      const exprt &struct_offset=op.op1();
+      const exprt &struct_offset = to_byte_extract_expr(op).offset();
       exprt member_offset = from_integer(*offset_int, struct_offset.type());
       plus_exprt final_offset(struct_offset, member_offset);
       simplify_node(final_offset);
 
-      byte_extract_exprt result(op.id(), op.op0(), final_offset, expr.type());
+      byte_extract_exprt result(
+        op.id(), to_byte_extract_expr(op).op(), final_offset, expr.type());
```
Comment (tautschnig) [inline src/util/simplify_expr_struct.cpp]: Maybe we can avoid doing `to_byte_extract_expr` multiple times?

File: src/util/simplify_expr_struct.cpp [hunk 5]
```diff
@@ -206,10 +207,10 @@ simplify_exprt::simplify_member(const member_exprt &expr)
   {
     // Try to look through member(cast(x)) if the cast is between structurally
     // identical types:
-    if(op_type == op.op0().type())
+    if(op_type == to_typecast_expr(op).op().type())
     {
       auto new_expr = expr;
-      new_expr.struct_op() = op.op0();
+      new_expr.struct_op() = to_typecast_expr(op).op();
```
Comment (tautschnig) [inline src/util/simplify_expr_struct.cpp]: We should not do `to_typecast_expr` multiple times.

**Review by tautschnig — CHANGES_REQUESTED** [review]
Yes, but we should avoid redundant checks in multiple places.

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:555]: done

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/util/simplify_expr.cpp]: done

File: src/util/simplify_expr_array.cpp (hunk 3 above)
Comment (kroening) [reply src/util/simplify_expr_array.cpp]: done

File: src/util/simplify_expr_struct.cpp (hunk 4 above)
Comment (kroening) [reply src/util/simplify_expr_struct.cpp]: done

File: src/util/simplify_expr_struct.cpp (hunk 5 above)
Comment (kroening) [reply src/util/simplify_expr_struct.cpp]: done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c4240b45e4c3422488c6d51f75ccc840236543da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121883183

**Review by tautschnig — APPROVED** [review]

**Pull request merged**